### PR TITLE
Add spectral normalization for braxlines

### PR DESCRIPTION
**Pull request by frt03**
I added a spectral normalization feature for braxlines.

Spectral normalization improves the discriminator in [VGCRL](https://arxiv.org/abs/2106.01404). My implementation seems to work successfully in DIAYN/Ant.

**w/o SN**
- Training Curves
image
- Skills visualization
image

**w/ SN**
- Training Curves
image
- Skills visualization
image

### CODE REVIEW

**Comment by frt03** [comment]
@googlebot I signed it!

File: brax/training/networks.py [hunk 1]
```diff
@@ -52,24 +54,58 @@ def __call__(self, data: jnp.ndarray):
...
   Args:
     layer_sizes: layers
     obs_size: size of an observation
     activation: activation
+    spectral_norm: whether to use a spectral normalization (default: False).
 
   Returns:
     a model
   """
   module = MLP(layer_sizes=layer_sizes, activation=activation)
```
Comment (erikfrey) [inline brax/training/networks.py:96]: You can delete this line.

File: brax/training/spectral_norm.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+"""Flax-style Dense module with Spectral Normalization.
+
+Reference:
+  Dense: https://github.com/google/flax/blob/main/flax/linen/linear.py
+  Spectral Normalization:
+    - https://arxiv.org/abs/1802.05957
+    - https://github.com/deepmind/dm-haiku/blob/main/haiku/_src/spectral_norm.py
+"""
+from typing import (Any, Callable, Tuple)
```
Comment (erikfrey) [inline brax/training/spectral_norm.py:23]: nit: remove the parentheses to follow the style of the rest of the codebase

File: brax/training/spectral_norm.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+
+default_kernel_init = lecun_normal()
+
+PRNGKey = Any
+Array = Any
+Shape = Tuple[int]
+Dtype = Any
+
+_no_init = lambda rng, shape: ()
```
Comment (erikfrey) [inline brax/training/spectral_norm.py:39]: nit: this seems to be unused?

File: brax/training/spectral_norm.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+from typing import (Any, Callable, Tuple)
+
+from jax import lax
+import jax.numpy as jnp
+
+from flax import linen
+from flax.linen.initializers import lecun_normal, normal, zeros
+
+
+default_kernel_init = lecun_normal()
```
Comment (erikfrey) [inline brax/training/spectral_norm.py:32]: style nit: module level constants should be all-caps: `DEFAULT_KERNEL_INIT`

that said, since you are only using this in one place, I think the code is a bit more readable if you just do this below:

```
kernel_init: Callable[[PRNGKey, Shape, Dtype], Array] = lecun_normal()
```

File: brax/training/spectral_norm.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                        self.kernel_init,
+                        (inputs.shape[-1], self.features))
+    kernel = jnp.asarray(kernel, self.dtype)
+
+    """for spectral normalization"""
+    kernel_shape = kernel.shape
+    # Handle scalars.
+    if kernel.ndim <= 1:
+      raise ValueError("Spectral normalization is not well defined for "
+                      "scalar inputs.")
```
Comment (erikfrey) [inline brax/training/spectral_norm.py:103]: nit: spacing, move right one space

File: brax/training/spectral_norm.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    u0_state.value = u0
+
+    y = lax.dot_general(inputs, kernel,
+                        (((inputs.ndim - 1,), (0,)), ((), ())),
+                        precision=self.precision)
+    if self.use_bias:
+      bias = self.param('bias', self.bias_init, (self.features,))
+      bias = jnp.asarray(bias, self.dtype)
+      y = y + bias
+    return y
```
Comment (erikfrey) [inline brax/training/spectral_norm.py:133]: nit: missing newline

**Review by erikfrey — CHANGES_REQUESTED** [review]
This is great!  Really nice clean code - just some style nits and then we are good to go.

**Comment by frt03** [comment]
@erikfrey I reflected your suggestions, thanks!
